### DISCUSSION TITLE
fileprovider fails to open file with write permission

### DISCUSSION THREAD

**Discussion by guidoDimasi** [dotnet/maui#26035] [Q&A]
### Description

hi, i'm building an app that download and store file inside the android file system (into a subfolder of /documents).
after downloading it i need to open selected file with the best aviable app (ex: if a file has .xlsx extension it should open with excel app) and into the app it should be able to modify and save in the same path.
the problem is that if my application target the sdk v33 the file opens but it remanin in what it seems read-only mode because if the user try to save the changes excel promt to the "save as" tab.
instead if my application target sdk less then 24 i could open the file without using the file provider (and it opens with write permission and i can save the changes).

whta i'm missing to use the 33sdk and the file provider?

### Steps to Reproduce

1. Crate maui app

2. add this code to the androidmanifest:

```
<application android:allowBackup="true" android:icon="@mipmap/appicon" android:supportsRtl="true" android:usesCleartextTraffic="true" android:label="DimasiGestioneDocumentale">
<provider android:name="androidx.core.content.FileProvider" android:authorities="${applicationId}.provider" android:exported="false" android:grantUriPermissions="true">
<meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/provider_path" />
</provider>
</application>
```

3. add the provider path file inside the android/resources/xml with this text:
```
<paths xmlns:android="http://schemas.android.com/apk/res/android">
<external-path name="external_files" path="."/>
</paths>
```

4. add a custom android service with the method:
```
public async void OpenFile(FileInfo fileInfo)
{
         Java.IO.File myDir = new(fileInfo.Directory.FullName);
         myDir.Mkdir();
         Java.IO.File file = new(myDir, fileInfo.Name);
         file.SetReadable(true);
         file.SetExecutable(true);
         file.SetWritable(true);
         if (!file.Exists()) return;
         if (!file.CanRead()) return;
         if (!file.CanWrite()) return;
 
         var fileUri = CustomFileProvider.GetUriForFile(Android.App.Application.Context, Android.App.Application.Context.PackageName + ".provider", file);
 
         fileUri.BuildUpon().AppendQueryParameter("grant_read_uri_permission", "true").Build();
         fileUri.BuildUpon().AppendQueryParameter("grant_write_uri_permission", "true").Build();
 
         var intent = new Intent(Intent.ActionView);
         intent.AddFlags(ActivityFlags.NewTask);
         intent.AddFlags(ActivityFlags.GrantReadUriPermission | ActivityFlags.GrantWriteUriPermission);
 
         string extension = MimeTypeMap.GetFileExtensionFromUrl(fileInfo.Name);
         string mimeType = GetMimeType(extension);

         intent.SetDataAndType(fileUri, mimeType);
         Android.App.Application.Context.StartActivity(intent);
}
```

5. download the file, save it in a folder like "/storage/emulated/0/Documents/MyAppaDocuments/Asset/TestExcel.xlsx" with succes
6. pass the fileinfo to the android service

the file opens but it seems readonly

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

Android

### Affected platform versions

android sdk 33

### Did you find any workaround?

there is a workaround:
1. target the sdk v23
2. pass the file path without using the FileProvider

intent.SetDataAndType(Android.Net.Uri.FromFile(file), mimeType);

**Comment by SupermindPT** [comment]
Hi there. Android 11 introduced changes to how applications handle files. Applications have a reserved directory where they can create or read files. If you want to read or open files outside of the reserved directory, you need to declare the MANAGE_ALL_FILES permission in the manifest.

Read more here: 
[MANAGE_ALL_FILES permission](https://developer.android.com/training/data-storage/manage-all-files )
[Android 11 scoped storage](https://source.android.com/docs/core/storage/scoped#:~:text=The%20goal%20of%20scoped%20storage%20is%20to%20protect,user%20documents%20downloaded%20to%20Download%20or%20other%20folders.)

**Comment by guidoDimasi** [comment]
my application already declare MANAGE_ALL_FILES, in fact the application write file inside the shared /documents folder of the system (the one reachable from every app, i can use a folder explorer to view the file even without my app)
the problem is that i wanna launch excel (or any other app for the desired extension)  and have the ability to modify and save the file.

**Comment by guidoDimasi** [comment]
i've tagged this issue as a bug because i think i'm using the right Intent command and maybe there is something inside MAUI that doesnt work as expected. if not i could use some help to refind the Intent command.

**Comment by SupermindPT** [comment]
Typically you would use the Launcher API to launch the apropriate app for the file extension.

If the problem is that you can't save the changes from the external Excel app that is not related to MAUI.

**Comment by guidoDimasi** [comment]
the laouncher API documentation for open file with third party app write:

`await Launcher.Default.OpenAsync(new OpenFileRequest(popoverTitle, new ReadOnlyFile(file)));`

so i avoided it because i need to open and edit the file.
without the launcher my solution was to write an intent following android documentation

**Comment by SupermindPT** [comment]
I would search the documentation of the Excel app you are using. Maybe you need to invoke some Uri.

It is very likely newer versions of Android are causing what you describe due to new security restrictions.

**Comment by guidoDimasi** [comment]
i've follow many online resoruces, the file uri seems correct "content://my.software.app/documents/file.xlsx"  (in fact the file can be viewed).
maybe there is something with manifest, the file provider or the intent command that on maui need to be wrtitten in some specific way.
